vowpalwabbit/offset_tree.cc

```diff
@@ -0,0 +1,296 @@
+      for (uint32_t i = 0; i < _num_leaf_nodes; i++)
+      {
+        nodes.emplace_back(i, 0, 0, 0, true);
+        tournaments.emplace_back(i);
+      }
+      while (tournaments.size() > 1)
+      {
+        // ids of next set of nodes start with the current
+        const auto num_tournaments = tournaments.size();
+        // node count (since ids are zero based)
+        auto id = nodes.size();
+        std::vector<uint32_t> new_tournaments;
+        new_tournaments.reserve(num_tournaments / 2 + 1);
+        for (size_t j = 0; j < num_tournaments / 2; ++j)
+        {
+          auto left = tournaments[2 * j];
+          auto right = tournaments[2 * j + 1];
+          new_tournaments.emplace_back(id);
+          nodes[left].parent_id = id;
+          nodes[right].parent_id = id;
+          nodes.emplace_back(id++, left, right, 0, false);
+        }
+        if (num_tournaments % 2 == 1)
+          new_tournaments.emplace_back(tournaments.back());
+        tournaments = std::move(new_tournaments);
+      }
+      root_idx = tournaments[0];
+      nodes[nodes.size()-1].parent_id = root_idx;
+    }
+    catch (std::bad_alloc& e)
+    {
+      THROW("Unable to allocate memory for offset tree.  Label count:" << _num_leaf_nodes << " bad_alloc:" << e.what());
+    }
+    _initialized = true;
+  }
+
+  uint32_t min_depth_binary_tree::internal_node_count() const { return nodes.size() - _num_leaf_nodes; }
+
+  uint32_t min_depth_binary_tree::leaf_node_count() const { return _num_leaf_nodes; }
+
+  void offset_tree::init(uint32_t num_actions) { binary_tree.build_tree(num_actions); }
+
+  int32_t offset_tree::learner_count() const { return binary_tree.internal_node_count(); }
+
+  // Helper to deal with collections that don't start with an index of 0
+  template<typename T>
+  struct offset_helper
+  {
+    // typedef verbose prediction buffer type
+    offset_helper(T& b, uint32_t index_offset)
+    : start_index_offset{index_offset}, collection(b) {}
+
+    // intercept index operator to adjust the offset before
+    // passing to underlying collection
+    typename T::const_reference operator[](size_t idx) const
+    {
+      return collection[idx - start_index_offset];
+    }
+
+    typename T::reference operator[](size_t idx) {
+      return collection[idx - start_index_offset];
+    }
+
+  private:
+    uint32_t start_index_offset = 0;
+    T& collection;
+  };
+
+  const offset_tree::scores_t& offset_tree::predict(LEARNER::single_learner& base, example& ec)
+  {
+    // - pair<float,float> stores the scores for left and right nodes
+    // - prediction_buffer stores predictions for all the nodes in the tree for the duration
+    //   of the predict() call
+    // - static thread_local ensures only one copy per calling thread.  This is to reduce
+    //   memory allocations in steady state.
+    using predict_buffer_t = std::vector<std::pair<float, float>>;
+    static thread_local predict_buffer_t prediction_buffer(learner_count());
```

Comment:
Didn't we decide a while back to not use thread_local due to issues on some platforms?

Reply:
Yes.  Fixed.